### PR TITLE
Do not crash trying to read a not known file

### PR DESCRIPTION
**Pull request by dgdavid**
## Problem

When running the installation or upgrade reading repositories from disk partition via `install=` boot param, it could crash with a `TypeError` exception when the `device` parameter is not added to the "ZyppRepoURL". Thus, booting with `install=disk:/path/to/installation/media` instead of `install=disk:/path/to/installation/media?device=sdb1` will end up in a crashed installation because _libzypp does not know which partition on the disk should be used_

Screenshot_openSUSE-Leap_2021-07-02_16:33:22

* Trello (internal link): https://trello.com/c/SF91Redx
* Bugzilla: https://bugzilla.suse.com/show_bug.cgi?id=1186387

## Solution

Ideally, instead of crashing, it would be nice to display a user-friendly message telling what was wrong. However, the exception in the Ruby side is too generic to be aware that the `#<TypeError: no implicit conversion of nil into String>` crash happens exactly because a missing `device` param.

So, since `RepomdDownloader#primary_xmls` could return an empty list of path, the implemented solution is to ignore repos for which `Yast::Pkg.SourceProvideFile` returns nil. And, apparently, it works.

## Pending

- [ ] Unit tests (help needed)
- [ ] Bump version and changelog
- [ ] Once merged, apply same changes to SLE-15-SP3 and master branches

## Manual test

To check manually that this change works, below steps were given

* Install a SLE-15-SP2 in a virtual machine
* Add a new disk to the virtual machine
* Create two partitions in the new disk
* Create the _sp2_ and _sp3_ directories in the second partition, /dev/vdb2
* Extract an SLE-15-SP2 media in sp2
* Extract an online SLE-15-SP3 media in sp3
* Boot with the SLE-15-SP2 cd
* Select Upgrade with `install=disk:/sp3` boot option
* Run the migration.
* Migrate the system
* Check that system was migrated

## `disk:/` vs `hd:/`

During manual tests, I could check that the problem only appears when using _install=**disk**:/whatever_ but not with _install=**hd**:/whatever_. For the latter, "ZyppRepoURL" gets the `device=` parameter _automagically_, as well as it happens with "RepoURL". See below screenshot

<details>
<summary>Click to show/hide the screenshot</summary>

---

Screenshot_openSUSE-Leap_2021-07-06_09:16:34

</details>

I have been digging in the Linuxrc code and maybe this difference could be _killed_ in [url_print_zypp](https://github.com/openSUSE/linuxrc/blob/9346016b26ba883931aa78ba857bb22bbab14894/url.c#L1006), checking [if url schema is `inst_hd` **or `inst_disk`**](https://github.com/openSUSE/linuxrc/blob/9346016b26ba883931aa78ba857bb22bbab14894/url.c#L1090-L1094)

### CODE REVIEW

**Comment by wfeldt** [comment]
> I have been digging in the Linuxrc code and maybe this difference could be killed in url_print_zypp, checking if url schema is inst_hd or inst_disk

No, that's not the way. The difference between `disk` and `hd` is that `disk` applies to both `hd` and `cdrom`. linuxrc determines which it is on first usage and tracks the result.

`url_print_zypp` already checks this here:

https://github.com/openSUSE/linuxrc/blob/9346016b26ba883931aa78ba857bb22bbab14894/url.c#L1038-L1040

The bug is to not continue using `scheme` from there on. You should change `url->scheme` to `scheme` ini the code after the cited condition. This should solve the issue.

**Comment by dgdavid** [comment]
> 
> No, that's not the way. The difference between `disk` and `hd` is that `disk` applies to both `hd` and `cdrom`. linuxrc determines which it is on first usage and tracks the result.
> 
> `url_print_zypp` already checks this here:
> 
> https://github.com/openSUSE/linuxrc/blob/9346016b26ba883931aa78ba857bb22bbab14894/url.c#L1038-L1040
> 
> The bug is to not continue using `scheme` from there on. You should change `url->scheme` to `scheme` in the code after the cited condition. This should solve the issue.

Thanks a lot for the explanation.

**Review by lslezak — COMMENTED** [review]
Error reporting

File: src/lib/y2packager/repomd_downloader.rb [hunk 1]
```diff
@@ -78,10 +78,12 @@ def primary_xmls(force = false)
 
       # add a temporary repository for downloading the files via libzypp
       src = Yast::Pkg.RepositoryAdd("base_urls" => [url])
-      repos.map do |(_name, dir)|
+      xml_paths = repos.map do |(_name, dir)|
         # download the repository index file (repomd.xml)
         repomd_file = Yast::Pkg.SourceProvideFile(src, 1, File.join(dir, "repodata/repomd.xml"))
 
+        next unless repomd_file
```
[the remark this answers is not in the document]
Comment (dgdavid) [reply src/lib/y2packager/repomd_downloader.rb:85]: > I think it should abort after displaying this error as the user would get completely different result than expected.

:thinking: :thinking:  How can we directly abort from here?

[the remark this answers is not in the document]
Comment (lslezak) [reply src/lib/y2packager/repomd_downloader.rb]: Ah, I see... It's not trivial...

The description says `Returns an empty list if the URL or the repository is not valid`. So check the caller.
Or we could raise a specific exception here.

[the remark this answers is not in the document]
Comment (dgdavid) [reply src/lib/y2packager/repomd_downloader.rb]: I still finding it a bit complicated. 

[`Y2Packager::RepomdDownloader#primary_xmls`](https://github.com/yast/yast-packager/blob/4c2f9c53f9f4cb3b8efa648f086696d4c893f08d/src/lib/y2packager/repomd_downloader.rb#L67) is called by [ `Y2Packager::ProductLocation.scan`](https://github.com/yast/yast-packager/blob/4c2f9c53f9f4cb3b8efa648f086696d4c893f08d/src/lib/y2packager/product_location.rb#L63) which, as far as I could check, is called from 

*  [`Yast::PackagerRepositoriesIncludeInclude.createSourceImpl`](https://github.com/yast/yast-packager/blob/4c2f9c53f9f4cb3b8efa648f086696d4c893f08d/src/include/packager/repositories_include.rb#L65) / [`Yast::PackagerRepositoriesIncludeInclude.scan_products`](https://github.com/yast/yast-packager/blob/4c2f9c53f9f4cb3b8efa648f086696d4c893f08d/src/include/packager/repositories_include.rb#L358)
* [`Y2Packager::MediumType.detect_medium_type`](https://github.com/yast/yast-packager/blob/4c2f9c53f9f4cb3b8efa648f086696d4c893f08d/src/lib/y2packager/medium_type.rb#L129)

At first glance, looks like [`Yast::PackagerRepositoriesIncludeInclude.createSourceImpl`](https://github.com/yast/yast-packager/blob/4c2f9c53f9f4cb3b8efa648f086696d4c893f08d/src/include/packager/repositories_include.rb#L55-L64) _could_ return `:abort`, which is documented as _problematic_

[the remark this answers is not in the document]
Comment (dgdavid) [reply src/lib/y2packager/repomd_downloader.rb]: Anyway, looks like [`Y2Packager::MediumType.detect_medium_type`](https://github.com/yast/yast-packager/blob/4c2f9c53f9f4cb3b8efa648f086696d4c893f08d/src/lib/y2packager/medium_type.rb#L129) is the entry point. Returning an `:abort` from there _works_, but not sure why.

| Yast::Pkg.LastError | Popup shown when #detect_medium_type returns `:abort` |
| - | - |
| Yast::Pkg.LastError | Yast::Pkg.LastError |

**Comment by dgdavid** [comment]
Since card was already archived, let's close the PR too.

**Pull request closed without merging**